Aggregations using "missing" lead to AggregationExecutionException because they use wrong datatype if the field was never stored in an index

**Elasticsearch version**: 2.1.2 and 2.3.5

**Plugins installed**: [none]

**JVM version**:1.8.0_80

**OS version**:Windows 10

**Description of the problem including expected versus actual behavior**:

We are working on a feature where different fields can be added to documents and we want them to be handled as double value. 

However we cannot say upfront which datatype these items should have and thus cannot define datatypes directly, but rather need to use dynamic templates in the mapping, so our mapping is something like the following:

```
{
    "testtype": {
        "dynamic_templates": [{
            "num_fields": {
                "match": "amount*",
                "mapping": {
                    "type": "double"
                }
            }
        }]
    }
}
```

When using aggregations together with specifying a missing-value, we found that the returned aggregation-key is different depending upon if the field was ever stored in the index or not.

Also if we query across two indices, one which contains a value for that field and one which does not, we get the following exceptions when doing a Bucket-Term-Aggregation together with a missing-attribute:

> AggregationExecutionException[Merging/Reducing the aggregations failed when computing the aggregation [ Name: testagg, Type: terms ] because: the field you gave in the aggregation query existed as two different types in two different indices]

In the test-case below, the 2nd query returns a string "-1", although ideally we would get a Double value -1.0. In the 3rd query we get the exception.

**Steps to reproduce**:
1. See test-case below

```
public class OldIndexWithoutMappingTest extends ESIntegTestCase {
    @Test
    public void testTwoIndicesOneWithoutMapping() throws Exception {
        initializeIndex();

        // aggregate by terms with a missing-value set
        final TermsBuilder agg = AggregationBuilders.terms("testagg").field("amount-high").missing(-1);

        // first try with first index only, here the "missing" value is returned as double
        SearchResponse searchResponse = client().prepareSearch("testindex-1").addAggregation(agg).execute().actionGet();
        assertEquals(0, searchResponse.getFailedShards());
        assertEquals(2, searchResponse.getHits().getTotalHits());
        assertEquals(2, ((Terms)searchResponse.getAggregations().get("testagg")).getBuckets().size());
        assertEquals(1234.0, ((Terms)searchResponse.getAggregations().get("testagg")).getBucketByKey("1234.0").getKey());
        assertEquals(-1.0, ((Terms)searchResponse.getAggregations().get("testagg")).getBucketByKey("-1.0").getKey());

        // then with second index only, here the missing value is returned as string!
        searchResponse = client().prepareSearch("testindex-2").addAggregation(agg).execute().actionGet();
        assertEquals(0, searchResponse.getFailedShards());
        assertEquals(1, searchResponse.getHits().getTotalHits());
        assertEquals(1, ((Terms)searchResponse.getAggregations().get("testagg")).getBuckets().size());
        assertEquals("-1", ((Terms)searchResponse.getAggregations().get("testagg")).getBucketByKey("-1").getKey());

        // and finally with both indices, here it fails because we take the missing value from both indices and these are different!
        searchResponse = client().prepareSearch("testindex*").addAggregation(agg).execute().actionGet();
        assertEquals(0, searchResponse.getFailedShards());
        assertEquals(2, searchResponse.getHits().getTotalHits());
        assertEquals(2, ((Terms)searchResponse.getAggregations().get("testagg")).getBuckets().size());
    }

    private void initializeIndex() throws Exception {
        String testMapping = getMapping(XContentFactory.jsonBuilder().startObject().startObject("testtype")).string();
        System.out.println("Mapping: " + testMapping);

        assertAcked(prepareCreate("testindex-1").addMapping("testtype", testMapping));
        assertAcked(prepareCreate("testindex-2").addMapping("testtype", testMapping));

        List<IndexRequestBuilder> indexBuilders = new ArrayList<>();
        indexBuilders.add(client().prepareIndex("testindex-1", "testtype", "index1-1").setSource("{\"amount-high\":1234}"));
        indexBuilders.add(client().prepareIndex("testindex-1", "testtype", "index1-2").setSource("{}"));
        indexBuilders.add(client().prepareIndex("testindex-2", "testtype", "index2-1").setSource("{}"));
        indexRandom(true, indexBuilders);
    }

    private XContentBuilder getMapping(XContentBuilder testtype) throws IOException {
        return testtype.startArray("dynamic_templates")
                    .startObject()
                        .startObject("num_fields")
                            .field("match", "amount*")
                            .startObject("mapping")
                                .field("type", "double").endObject()
                        .endObject()
                    .endObject()
                .endArray();
    }
}
```

The strange thing here is that filtering/querying works fine, also aggregations work fine unless the missing-attribute is specified.

The full stacktrace is:
```
Failed to execute phase [merge], [reduce] 

	at __randomizedtesting.SeedInfo.seed([3052F79434FAD641:8EABC959AFC832B0]:0)
	at org.elasticsearch.action.search.SearchDfsQueryThenFetchAsyncAction$3.onFailure(SearchDfsQueryThenFetchAsyncAction.java:212)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: AggregationExecutionException[Merging/Reducing the aggregations failed when computing the aggregation [ Name: testagg, Type: terms ] because: the field you gave in the aggregation query existed as two different types in two different indices]
	at org.elasticsearch.search.aggregations.bucket.terms.InternalTerms.doReduce(InternalTerms.java:188)
	at org.elasticsearch.search.aggregations.InternalAggregation.reduce(InternalAggregation.java:153)
	at org.elasticsearch.search.aggregations.InternalAggregations.reduce(InternalAggregations.java:170)
	at org.elasticsearch.search.controller.SearchPhaseController.merge(SearchPhaseController.java:411)
	at org.elasticsearch.action.search.SearchDfsQueryThenFetchAsyncAction$3.doRun(SearchDfsQueryThenFetchAsyncAction.java:198)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	... 3 more
```

--- a/core/src/main/java/org/elasticsearch/search/aggregations/support/MissingValues.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/support/MissingValues.java
@@ -80,7 +80,7 @@ public enum MissingValues {
     }
 
     public static ValuesSource.Numeric replaceMissing(final ValuesSource.Numeric valuesSource, final Number missing) {
-        final boolean missingIsFloat = missing.longValue() != (long) missing.doubleValue();
+        final boolean missingIsFloat = missing.doubleValue() % 1 != 0;
         final boolean isFloatingPoint = valuesSource.isFloatingPoint() || missingIsFloat;
         return new ValuesSource.Numeric() {
 

--- a/core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParserHelper.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParserHelper.java
@@ -67,6 +67,17 @@ public final class ValuesSourceParserHelper {
         objectParser.declareField(ValuesSourceAggregationBuilder::missing, XContentParser::objectText,
                 new ParseField("missing"), ObjectParser.ValueType.VALUE);
 
+        objectParser.declareField(ValuesSourceAggregationBuilder::valueType, p -> {
+            ValueType valueType = ValueType.resolveForScript(p.text());
+            if (targetValueType != null && valueType.isNotA(targetValueType)) {
+                throw new ParsingException(p.getTokenLocation(),
+                        "Aggregation [" + objectParser.getName() + "] was configured with an incompatible value type ["
+                                + valueType + "]. It can only work on value of type ["
+                                + targetValueType + "]");
+            }
+            return valueType;
+        }, new ParseField("value_type", "valueType"), ObjectParser.ValueType.STRING);
+
         if (formattable) {
             objectParser.declareField(ValuesSourceAggregationBuilder::format, XContentParser::text,
                     new ParseField("format"), ObjectParser.ValueType.STRING);
@@ -75,17 +86,6 @@ public final class ValuesSourceParserHelper {
         if (scriptable) {
             objectParser.declareField(ValuesSourceAggregationBuilder::script, org.elasticsearch.script.Script::parse,
                     Script.SCRIPT_PARSE_FIELD, ObjectParser.ValueType.OBJECT_OR_STRING);
-
-            objectParser.declareField(ValuesSourceAggregationBuilder::valueType, p -> {
-                ValueType valueType = ValueType.resolveForScript(p.text());
-                if (targetValueType != null && valueType.isNotA(targetValueType)) {
-                    throw new ParsingException(p.getTokenLocation(),
-                            "Aggregation [" + objectParser.getName() + "] was configured with an incompatible value type ["
-                                    + valueType + "]. It can only work on value of type ["
-                                    + targetValueType + "]");
-                }
-                return valueType;
-            }, new ParseField("value_type", "valueType"), ObjectParser.ValueType.STRING);
         }
 
         if (timezoneAware) {

--- a/core/src/test/java/org/elasticsearch/search/aggregations/support/MissingValuesTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/support/MissingValuesTests.java
@@ -293,4 +293,9 @@ public class MissingValuesTests extends ESTestCase {
             }
         }
     }
+
+    public void testFloatingPointDetection() {
+        assertFalse(MissingValues.replaceMissing(ValuesSource.Numeric.EMPTY, 3).isFloatingPoint());
+        assertTrue(MissingValues.replaceMissing(ValuesSource.Numeric.EMPTY, 3.5).isFloatingPoint());
+    }
 }

--- a/docs/reference/aggregations.asciidoc
+++ b/docs/reference/aggregations.asciidoc
@@ -92,11 +92,15 @@ enables defining all the "dynamic" expressions in the script as parameters, whic
 between calls (this will ensure the use of the cached compiled scripts in Elasticsearch).
 ===============================
 
-Scripts can generate a single value or multiple values per document. When generating multiple values, one can use the
-`script_values_sorted` settings to indicate whether these values are sorted or not. Internally, Elasticsearch can
-perform optimizations when dealing with sorted values (for example, with the `min` aggregations, knowing the values are
-sorted, Elasticsearch will skip the iterations over all the values and rely on the first value in the list to be the
-minimum value among all other values associated with the same document).
+Elasticsearch uses the type of the field in the mapping in order to figure out
+how to run the aggregation and format the response. However there are two cases
+in which Elasticsearch cannot figure out this information: unmapped fields (for
+instance in the case of a search request across multiple indices, and only some
+of them have a mapping for the field) and pure scripts. For those cases, it is
+possible to give Elasticsearch a hint using the `value_type` option, which
+accepts the following values: `string`, `long` (works for all integer types),
+`double` (works for all decimal types like `float` or `scaled_float`), `date`,
+`ip` and `boolean`.
 
 --
 

--- a/rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/20_terms.yaml
+++ b/rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/20_terms.yaml
@@ -465,3 +465,132 @@ setup:
   - match: { aggregations.str_terms.buckets.0.key: 1234 }
 
   - match: { aggregations.str_terms.buckets.0.doc_count: 2 }
+
+---
+"Unmapped strings":
+
+  - do:
+      index:
+        index: test_1
+        type: test
+        id: 1
+        body: {}
+
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        body: { "size" : 0, "aggs" : { "string_terms" : { "terms" : { "field" : "unmapped_string", "value_type" : "string", "missing": "abc" } } } }
+
+  - match: { hits.total: 1 }
+
+  - length: { aggregations.string_terms.buckets: 1 }
+
+  - match: { aggregations.string_terms.buckets.0.key: "abc" }
+
+  - match: { aggregations.string_terms.buckets.0.doc_count: 1 }
+
+---
+"Unmapped booleans":
+
+  - do:
+      index:
+        index: test_1
+        type: test
+        id: 1
+        body: {}
+
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        body: { "size" : 0, "aggs" : { "boolean_terms" : { "terms" : { "field" : "unmapped_boolean", "value_type" : "boolean", "missing": true } } } }
+
+  - match: { hits.total: 1 }
+
+  - length: { aggregations.boolean_terms.buckets: 1 }
+
+  - match: { aggregations.boolean_terms.buckets.0.key: 1 }
+
+  - match: { aggregations.boolean_terms.buckets.0.key_as_string: "true" }
+
+  - match: { aggregations.boolean_terms.buckets.0.doc_count: 1 }
+
+---
+"Unmapped dates":
+
+  - do:
+      index:
+        index: test_1
+        type: test
+        id: 1
+        body: {}
+
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        body: { "size" : 0, "aggs" : { "date_terms" : { "terms" : { "field" : "unmapped_date", "value_type" : "date", "missing": "2016-05-11" } } } }
+
+  - match: { hits.total: 1 }
+
+  - length: { aggregations.date_terms.buckets: 1 }
+
+  - match: { aggregations.date_terms.buckets.0.key: 1462924800000 }
+
+  - match: { aggregations.date_terms.buckets.0.key_as_string: "2016-05-11T00:00:00.000Z" }
+
+  - match: { aggregations.date_terms.buckets.0.doc_count: 1 }
+
+---
+"Unmapped longs":
+
+  - do:
+      index:
+        index: test_1
+        type: test
+        id: 1
+        body: {}
+
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        body: { "size" : 0, "aggs" : { "long_terms" : { "terms" : { "field" : "unmapped_long", "value_type" : "long", "missing": 3 } } } }
+
+  - match: { hits.total: 1 }
+
+  - length: { aggregations.long_terms.buckets: 1 }
+
+  - match: { aggregations.long_terms.buckets.0.key: 3 }
+
+  - match: { aggregations.long_terms.buckets.0.doc_count: 1 }
+
+---
+"Unmapped doubles":
+
+  - do:
+      index:
+        index: test_1
+        type: test
+        id: 1 
+        body: {}
+
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        body: { "size" : 0, "aggs" : { "double_terms" : { "terms" : { "field" : "unmapped_double", "value_type" : "double", "missing": 3.5 } } } }
+
+  - match: { hits.total: 1 }
+
+  - length: { aggregations.double_terms.buckets: 1 }
+
+  - match: { aggregations.double_terms.buckets.0.key: 3.5 }
+
+  - match: { aggregations.double_terms.buckets.0.doc_count: 1 }

